File: src/js/esse/tests/validate.js
/* eslint-disable no-unused-expressions */
// eslint-disable-next-line import/no-extraneous-dependencies
import { expect } from "chai";
import file from "file";
import groupBy from "lodash/groupBy";
import path from "path";

import { ESSE } from "../index";
import { EXAMPLES_DIR, SCHEMAS_DIR } from "../settings";
import { parseIncludeReferenceStatements } from "../utils";

const esse = new ESSE();

describe("validate all examples", () => {
    file.walkSync(EXAMPLES_DIR, (dirPath_, dirs_, files_) => {
        files_.forEach((file_) => {
            const examplePath = path.join(dirPath_, file_);
            const schemaPath = examplePath.replace(EXAMPLES_DIR, SCHEMAS_DIR);
            if (!examplePath.endsWith(".json") || !schemaPath.endsWith(".json")) {
                // ignore files like .DS_Store
                return;
            }
            it(`${examplePath.replace(`${EXAMPLES_DIR}/`, "")}`, () => {
                const example = parseIncludeReferenceStatements(examplePath);
                const schema = parseIncludeReferenceStatements(schemaPath);
                const valid = esse.validate(example, schema);
                // eslint-disable-next-line no-unused-expressions
                expect(valid).to.be.ok;
            });
        });
    });
});

describe("schema titles must be unique or empty", () => {
    const repeatedSchemaTitles = Object.entries(groupBy(esse.schemas, "title"))
        .filter(([title, groupedValues]) => title !== "undefined" && groupedValues.length > 1)
        .map(([title, groupedValues]) => [title, groupedValues.map(({ $id }) => $id)]);

    console.log(repeatedSchemaTitles);

    expect(repeatedSchemaTitles).to.be.an("array").that.is.empty;
});

File: src/py/esse/tests/validate.py
import os
import unittest

from esse import ESSE
from esse.settings import SCHEMAS_DIR, EXAMPLES_DIR
from esse.utils import parseIncludeReferenceStatements

esse = ESSE()


class TestCase(unittest.TestCase):

    def __init__(self, example_path):
        super(TestCase, self).__init__(methodName="runTest")
        self.example_path = example_path
        self.schema_path = self.example_path.replace(EXAMPLES_DIR, SCHEMAS_DIR)

    def runTest(self):
        schema = parseIncludeReferenceStatements(self.schema_path)
        example = parseIncludeReferenceStatements(self.example_path)
        esse.validate(example, schema)


class TestResult(unittest.TextTestResult):
    def getDescription(self, test):
        return test.example_path.replace("".join((EXAMPLES_DIR, "/")), "")


def validate_examples():
    """
    Validates examples against their schemas.
    """
    suite = unittest.TestSuite()
    for root, dirs, files in os.walk(EXAMPLES_DIR):
        for file_ in files:
            TestCase(example_path=os.path.join(root, file_)).run()
            suite.addTest(TestCase(example_path=os.path.join(root, file_)))
    unittest.TextTestRunner(verbosity=2, resultclass=TestResult).run(suite)


if __name__ == "__main__":
    validate_examples()
